source/common/http/conn_manager_impl.cc

```diff
@@ -470,6 +487,47 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
       connection_manager_.config_, *snapped_route_config_, connection_manager_.random_generator_,
       connection_manager_.runtime_, connection_manager_.local_info_);
 
+  if (!cached_route_.valid()) {
```

Comment:
I don't think it's possible for it to be valid here? Can you switch this to an ASSERT followed by a direct set? Also, I think that means that elsewhere where route is fetched you can remove the guard and just return the value.